Better typing for Job and JobRunners

potiuk opened the pull request:
By leveraging Generics, the typing for Runners and Job and JobPydantic is now more complete and accurate.

* Scheduler and Backfill Runners limit their code to Job and can use all the things that ORM Job allows them to do

* Other runners are limited to union of Job and JobPydantic version so that they can be run on the client side of the internal API without having all the Job features.

This is a follow up after #31182 that fixed missing job_type for DagProcessor Job and nicely extracted job to BaseRunner but broke MyPy/Typing guards implemented in the runners that should aid the AIP-44 implementation.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

uranusjr commented on airflow/jobs/base_job_runner.py:
```diff
@@ -17,22 +17,27 @@
     from airflow.jobs.job import Job
+    from airflow.serialization.pydantic.job import JobPydantic
 
+T = TypeVar("T", bound="Job | JobPydantic")
```
```suggestion
    T = TypeVar("T", Job, JobPydantic)
```

I think this is equivalent and easier to read?

potiuk replied:
Nice!

uranusjr commented on airflow/jobs/base_job_runner.py:
```diff
@@ -17,22 +17,27 @@
+T = TypeVar("T", bound="Job | JobPydantic")
 
-class BaseJobRunner:
+JobOrJobPydantic = TypeVar("JobOrJobPydantic", bound="Job | JobPydantic")
```
Only one of this and `T` needs to exist.

potiuk replied:
Even nicer!

uranusjr commented on airflow/jobs/base_job_runner.py:
```diff
@@ -17,28 +17,31 @@
     from airflow.jobs.job import Job
+    from airflow.serialization.pydantic.job import JobPydantic
 
+JobOrJobPydantic = TypeVar("JobOrJobPydantic", Job, JobPydantic)
```
I think this needs to be in the `TYPE_CHECKING` block to avoid NameError at runtime

potiuk replied:
yep

uranusjr commented on airflow/jobs/base_job_runner.py:
```diff
@@ -17,28 +17,31 @@
                 f"This is a bug and should be reported."
             )
-        self.job = job
+        self.job: JobOrJobPydantic = job
```
I don’t think this annotation is needed?

potiuk replied:
MyPy complained when it was not there - not sure why

potiuk replied:
MyPy complained hen it wasn't there. But just wondering - isn't the original approach with both  T and JobOrJobPydantic needed actually? The first is the generic type (which can also be just Job), and the other actually a type definition?

uranusjr replied:
> The first is the generic type (which can also be just Job), and the other actually a type definition

A TypeVar is a, well, variable (name), and that variable name can be used in different contexts. One name can serve these two purposes without conflicting internally, so it works fine.

uranusjr replied:
> MyPy complained when it was not there - not sure why

What’s the error message? Maybe that can be categorised as a deficit and submitted to Mypy for fixing.

potiuk replied:
```
airflow/jobs/base_job_runner.py:46: error: Need type annotation for "job" 
[var-annotated]
            self.job = job
                       ^
```

potiuk replied:
Pretty straightforward :)

potiuk replied:
> A TypeVar is a, well, variable (name), and that variable name can be used in different contexts. One name can serve these two purposes without conflicting internally, so it works fine.

I see  - but I think (i just re-added it). it's a bit clearer and more explicit now what I separated them into two vars - It would be rather confusing (pairing with the mypy deficit) to see the job declared with JobOrPydantic, where it can be just Job actually. I think having T makes it  a bit clearer what the intention is here.

potiuk replied:
BTW. This is what I do not like too much about the typing system's "flexibility" in Python (especially when it is abused) - same things can be used in different context with different meaning and sometimes when I look at the code that has them mixed, it's really hard to wrap your head around it.

potiuk committed:
Better typing for Job and JobRunners

By avoiding setting the job in the BaseJobRunner, the typing for Runners
and Job and JobPydantic is now more complete and accurate.

Scheduler and Backfill Runners limit their code to Job and can use all
the things that ORM Job allows them to do

Other runners are limited to union of Job and JobPydantic version so
that they can be run on the client side of the internal API without
having all the Job features.

This is a follow up after #31182 that fixed missing job_type for
DagProcessor Job and nicely extracted job to BaseRunner but broke
MyPy/Typing guards implemented in the runners that should aid the AIP-44
implementation.

uranusjr reviewed (approved):
I’ll experiement some other approaches later as well, this assignment still fits better in the base class.
